dev-client/src/screens/AppScaffold.tsx

```diff
@@ -1,80 +1,137 @@
+};
 
-type ScreenMapArgs = [ScreenRoutes, ScreenConfig<keyof RootStackParamList>];
+type ScreenName = keyof typeof screens;
+type RootStackParamList = {
+  [K in ScreenName]: Parameters<(typeof screens)[K]['View']>[0];
+};
 
-const previews = fetchProjects();
+const RootStack = createNativeStackNavigator<RootStackParamList>();
+export type RootStackScreenProps = NativeStackScreenProps<
+  RootStackParamList,
+  ScreenName
+>;
 
-function mapScreens(t: TFunction) {
-  return ([name, config]: ScreenMapArgs) => {
-    // TODO: initialParams are stubs, a stopgap for while we are not connected to the backend
-    // This setup should be changed when we get to connecting the backend
-    let initialParams;
-    switch (name) {
-      case 'PROJECT_LIST':
-        initialParams = {projects: previews};
-        break;
-    }
+type HeaderTitleType<Props> = (
+  _: Props & React.ComponentProps<typeof NavigationHeaderTitle>,
+) => React.ReactNode;
+
+type RootStackNavigationOptions<Props> = Omit<
+  NativeStackNavigationOptions,
+  'headerTitle'
+> & {
+  HeaderTitle?: HeaderTitleType<Props>;
+};
 
-    let options: NativeStackNavigationOptions = {};
+export type ScreenDefinition<Props = undefined> = {
+  View: (T: Props) => React.ReactNode;
+  options?: (_: Props) => RootStackNavigationOptions<Props>;
+};
 
-    if (config.hideBack) {
-      options.headerBackVisible = false;
-    }
+const DefaultHeader = ({
+  name,
+  ...props
+}: React.ComponentProps<typeof NavigationHeaderTitle> & {name: string}) => {
+  const {t} = useTranslation();
+  return (
+    <NavigationHeaderTitle {...props}>
+      {t('screens.' + name) ?? name}
+    </NavigationHeaderTitle>
+  );
+};
 
-    options.headerLeft = config.leftIcon;
-    options.headerRight = config.rightIcon;
-
-    return (
-      <Stack.Screen
-        name={name}
-        component={config.screen}
-        initialParams={initialParams}
-        key={name}
-        options={({route}) => {
-          if (config.paramTitle !== undefined) {
-            options.headerTitle = config.paramTitle({t, route});
-          } else {
-            options.headerTitle = t('screens.' + name) ?? undefined;
-          }
-          return options;
-        }}
-      />
-    );
+const defineScreen = <
+  Name extends ScreenName,
+  Props extends RootStackParamList[Name],
+>(
+  name: Name,
+  {options, View}: ScreenDefinition<Props>,
+) => {
+  const wrappedOptions = (
+    props: RootStackScreenProps,
+  ): NativeStackNavigationOptions => {
+    const {HeaderTitle = undefined, ...origOptions} =
+      options === undefined ? {} : options(props.route.params as Props);
+    const CastTitle = HeaderTitle as HeaderTitleType<object> | undefined;
+    return {
+      ...origOptions,
+      headerTitle: ({tintColor}) =>
+        CastTitle !== undefined ? (
+          <CastTitle {...props.route.params} tintColor={tintColor} />
+        ) : (
+          <DefaultHeader name={name} tintColor={tintColor} />
+        ),
+    };
```

Comment:
question: OK, do you think there is a benefit to being able to define the header like this? The way it was implemented previously, any params for the title name could be passed to the screen as route params. We could just have the translation function do `t("screen.name", route.params.titleParams)`. I think this is cleaner than having the screen itself create its own name by looking at global state, especially if we want to have some sort deep linking eventually. I'm curious what you think!

Reply:
i think we need to be able to compute it from the redux state regardless because e.g. if the user changes the site's name from the site's dashboard it won't update in the header if the name was passed as a screen parameter. same if someone changes the name remotely and they get synced to the client while they are on a screen with a dynamically computed title

Reply:
OK good point. I agree, we can come back if this starts causing problems.